### PR TITLE
GH-127090: Fix `urllib.response.addinfourl.url` value for opened `file:` URIs

### PR DESCRIPTION
**Pull request by barneygale**
The canonical `file:` URL (as generated by `pathname2url()`) is now used as the `url` attribute of the returned `addinfourl` object. The `addinfourl.url` attribute reflects the resolved URL for both `file:` or `http[s]:` URLs now.

* Issue: gh-127090

### CODE REVIEW

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM in general, but there are not enough tests.

I think that you can add checks for `url` in `test_file_notexists` and maybe in `test_basic` and `test_copy` in `urlretrieve_FileTests`.

**Review by serhiy-storchaka — COMMENTED** [review]
On other hand, the current code removes port and fragment from url, but the changed code will preserve them. I do not know what is more correct.

For history of the current code see 18e4dd74e06a8d036e46336041a72f3dae52b15c and 6057ba1f97ea19c484b03d5dd2f431e7767fefd1 (bpo-8656/gh-52902).

**Comment by barneygale** [comment]
Thank you Serhiy, that's very helpful.

I'll mark this PR as a draft for now. Once #127125 and #127194 land, I'll adjust this patch to call `pathname2url()` to canonicalise the URL, which will strip the port and fragment, restoring previous behaviour. It looks like `urlopen()` canonicalises HTTP URLs, and I think it makes sense to do so here.

**Comment by barneygale** [comment]
> I think that you can add checks for `url` in `test_file_notexists` and maybe in `test_basic` and `test_copy` in `urlretrieve_FileTests`.

I've expanded `test_file_notexists`, but the `urlretrieve()` tests aren't relevant here - they use a different codepath for handling local files (with its own set of problems...)

> On other hand, the current code removes port and fragment from url, but the changed code will preserve them. I do not know what is more correct.

The existing code raises `URLError` if a port is given, and that's tested in `test_urllib2.HandlerTests.test_file`. I've slightly expanded that function to also test that fragments are stripped.

File: Lib/test/test_urllib2.py [hunk 1]
```diff
@@ -818,19 +810,20 @@ def test_file(self):
         o = h.parent = MockOpener()
 
         TESTFN = os_helper.TESTFN
-        urlpath = sanepathname2url(os.path.abspath(TESTFN))
         towrite = b"hello, world\n"
+        canonurl = 'file:' + urllib.request.pathname2url(os.path.abspath(TESTFN))
+        parsed = urlparse(canonurl)
```
Comment (serhiy-storchaka) [inline Lib/test/test_urllib2.py]: I think that `urlparse()` will be deprecated. It is better to use `urlsplit()`.

File: Lib/test/test_urllib2.py [hunk 2]
```diff
@@ -818,19 +810,20 @@ def test_file(self):
...
         urls = [
-            "file://localhost%s" % urlpath,
-            "file://%s" % urlpath,
-            "file://%s%s" % (socket.gethostbyname('localhost'), urlpath),
+            canonurl,
+            parsed._replace(netloc='localhost').geturl(),
```
Comment (serhiy-storchaka) [inline Lib/test/test_urllib2.py:820]: How would this work on Windows if the current directory is a share ("`\\server\share\...`")?

Comment (barneygale) [reply Lib/test/test_urllib2.py:820]: Thanks - the test is now skipped in that case.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Comment by serhiy-storchaka** [comment]
Shouldn't it be backported?

**Comment by picnixz** [comment]
On the issue, Barney decided that it was not worth backporting.

**Pull request merged**